File: AspNetMonsters.Blazor.Geolocation/src/Location.ts
declare var DotNet: any;

class coordinate {
    public latitude: number = 0;
    public longitude: number = 0;
    public accuracy: number = 0;
}


let assemblyName = "AspNetMonsters.Blazor.Geolocation";
let namespace = "AspNetMonsters.Blazor.Geolocation";
let type = "LocationService";
async function dispatchResponse(id: string, location: coordinate) {
    await DotNet.invokeMethodAsync(namespace, 'ReceiveResponse', id, location.latitude, location.longitude, location.accuracy);
}

async function dispatchWatchResponse(id: string, location: coordinate) {
    await DotNet.invokeMethodAsync(namespace, 'ReceiveWatchResponse', id, location.latitude, location.longitude, location.accuracy);
}

window['AspNetMonsters'] = window['AspNetMonsters'] || {};
window['AspNetMonsters']['Blazor'] = window['AspNetMonsters']['Blazor'] || {};
window['AspNetMonsters']['Blazor']['Geolocation'] = window['AspNetMonsters']['Blazor']['Geolocation'] || {};

window['AspNetMonsters']['Blazor']['Geolocation']['GetLocation'] = (requestId) => {
    if (navigator.geolocation) {
        navigator.geolocation.getCurrentPosition((position) => {
            dispatchResponse(requestId, position.coords);
        });
    }
    else {
        return "No location finding";
    }
};

window['AspNetMonsters']['Blazor']['Geolocation']['WatchLocation'] = (requestId) => {
    if (navigator.geolocation) {
        navigator.geolocation.watchPosition((position) => {
            dispatchWatchResponse(requestId, position.coords);
        });
    }
    else {
        return "No location watching";
    }
};

File: AspNetMonsters.Blazor.Geolocation/Location.cs
namespace AspNetMonsters.Blazor.Geolocation
{
    public class Location
    {
        public decimal Latitude { get; set; }
        public decimal Longitude { get; set; }
        public decimal? Altitude { get; set; }
        public decimal Accuracy { get; set; }
        public decimal? AltitudeAccuracy { get; set; }
        public decimal? Heading { get; set; }
        public decimal? Speed { get; set; }


        public override string ToString()
        {
            return $"Location: ({Latitude}, {Longitude}) with accuracy {Accuracy}";
        }
    }
}
